### ISSUE TITLE
review data-specific modifications for RECO related to HCAL noise flagging

### ISSUE THREAD

**Issue by slava77** [cms-sw/cmssw#27767]
We (historically) have several settings which are applied specifically for data processing (different from what is done for MC)
https://github.com/cms-sw/cmssw/blob/CMSSW_11_0_0_pre5/Configuration/StandardSequences/python/Reconstruction_Data_cff.py
To complicate this, for most of the processing eras these should be replaced with customisePostEra\* methods from https://github.com/cms-sw/cmssw/blob/CMSSW_11_0_0_pre5/Configuration/DataProcessing/python/RecoTLR.py#L47-L51

-  particleFlowRecHitHF modifications are applied to all data adding a cleaning for severity above 9 and threshold of 30 specifically for HFDigiTime flag. Compared to MC which has already cleaning for severity above 11 for all flags.
-  ```AddFlag(hcalRecAlgos,"HFDigiTime",11)``` compared to MC settings of 8 prior to 2017 (HCAL phase-0) and not known/ignored starting from 2017.
    - it is reset to a lower severity of 8 in 2015 25 ns and 2016 with ```AddFlag(process.hcalRecAlgos,"HFDigiTime",8)```.
    - it is removed as ``` RemoveFlag(process.hcalRecAlgos,"HFDigiTime")``` starting from 2017 via ```customisePostEra_Run2_2017```.  
-  ```AddFlag(hcalRecAlgos,"HBHEFlatNoise",12)``` compared to MC settings of 8
    - it is changed to lower severity of 8 via ```AddFlag(process.hcalRecAlgos,"HBHEFlatNoise",8)``` for all data starting from 2015 25 ns
    - PF HBHE hits are rejected with severity > 11
-  ```AddFlag(hcalRecAlgos,"HBHENegativeNoise",12)``` compared to MC settings of 8
    - PF HBHE hits are rejected with severity > 11

In all cases where settings in MC differ from data, the choice is generally assuming that MC does not produce real noise and should not be flagged.
This is despite the fact that if real signal looks like noise in bulk, its hits will be rejected in PF reco in real data.

One scenario where this approach is OK is for MC made before prompt reco and data behavior is unknown. In this case, if we do not know in advance if a filter rejects anything in data, it perhaps does not make sense to apply it on MC especially if it can lead to efficiency losses. It seems important in this case to have the hits flagged and still available at least in AOD (but better yet in miniAOD) in MC and in data.

Some examples of flagging of noise-like events in signal would be very useful to see if the current asymmetric approach (MC passes the hits, data rejects them) is more practical compared to a symmetric approach where both data and MC will behave the same.

The goal of this issue is for each case to see if the flag settings can be moved to the common area and apply for both data and MC.
Additionally, some clarification is needed if relevant data about cleaned hits is saved in miniAOD (or is it available in AOD).

Detailed discussion, should probably be done by email or HN/egroup. This issue is mostly to track the conclusions to see if anything can/should be done in CMSSW.

Timescale target: 11_X or later (aiming for Run 3 or/and if needed Run 2 software integration or potential reprocessing/rereco tests).
Perhaps this can be reviewed during the HCAL days in September.

@mariadalfonso @abdoulline @hatakeyamak @bendavid @lathomas 
@jaehyeok 
please check also if the issue correctly describes the settings of severities in parts were they differ between data and MC.
I'd be happy to replace it with a more authoritative source if there is one.

**Comment by slava77** [comment]
assign reconstruction

**Comment by abdoulline** [comment]
Hi Slava, all, 

first a small comment on your first bullet - just to avoid misunderstanding: 
"Compared to MC which has already cleaning for severity above 11 for all flags"
 - in MC thoughout all the modules there is accepted SeverityLevel=9, above which the hits are rejected. For HF flags in question assigned SeverityLevel is 11 (both for MC and data):
https://github.com/cms-sw/cmssw/blob/master/RecoLocalCalo/HcalRecAlgos/plugins/HcalRecAlgoESProducer.cc#L186
-> 
https://github.com/cms-sw/cmssw/blob/master/RecoLocalCalo/HcalRecAlgos/python/hcalRecAlgoESProd_cfi.py#L46

Now,  if we put aside  2018 HI (re-)reco config issue, which I've come across on Monday, and which we've discussed in a dedicated thread with you (with Andrea Perrotta, Jae, Igor and Chandi included),  for which you've submitted (already merged PR https://github.com/cms-sw/cmssw/pull/27755), 
*all* we have as a difference between MC and data in 2018 (for instance) is HBHENegativeNoise (applied to Phase0, so HB *only*, as HE is Phase1, not applied to MC), the rest is the same
between MC and data.

Now, looking specifically at this flag in RelVal plots:

(1) this particular bit (#27) [*] is set for JetHT 2018C (for instance)
https://tinyurl.com/y5fodl47
(probability per recorded HB RecHit < 10^-5)

(2) for MC it can be set (but not used due to low SeverityLevel),
but for instance in TTBar PU 2018 at the given statistics (9K events) there is none set:
https://tinyurl.com/y3gmuw99

So, looks like MC doesn't mimic (at this level of MC statistics) this particular bit in question.
Also, the way we look at it in MC (2) can be done for any MC of special interest (exotic signal which is susceptible to suffer from this particular Phase0 HBHE filter) in Run2  as I've mentioned earlier.
For Run3 is will not be active anyway. So I don't see the problem here.

[*]
https://cmssdt.cern.ch/lxr/source/DataFormats/METReco/interface/HcalPhase1FlagLabels.h

**Comment by abdoulline** [comment]
adding HCAL Noise convener Chandi  @cpkar

**Comment by abdoulline** [comment]
adding HCAL DPG convener Aram @arapyan

**Comment by slava77** [comment]
> first a small comment on your first bullet - just to avoid misunderstanding:
> "Compared to MC which has already cleaning for severity above 11 for all flags" :
> in MC throughout all the modules there is accepted SeverityLevel=9, above which the hits are rejected. Except for towerMakerPF , where level 11 is accepted:
> https://cmssdt.cern.ch/lxr/source/RecoParticleFlow/PFClusterProducer/python/towerMakerPF_cfi.py#0056
> 
> For HF flags in question assigned SeverityLevel is 11 (both for MC and data):
> https://github.com/cms-sw/cmssw/blob/master/RecoLocalCalo/HcalRecAlgos/plugins/HcalRecAlgoESProducer.cc#L186
> ->
> https://github.com/cms-sw/cmssw/blob/master/RecoLocalCalo/HcalRecAlgos/python/hcalRecAlgoESProd_cfi.py#L46

We do not use towerMakerPF in standard reco. My comment was about particleFlowRecHitHF, which has max severity of 11 for MC for the "Standard" flags (which includes all flags).
https://github.com/cms-sw/cmssw/blob/master/RecoParticleFlow/PFClusterProducer/python/particleFlowRecHitHF_cfi.py#L26-L29
The severity threshold of 9 is only for HFLongShort and HFLongShort flags 
https://github.com/cms-sw/cmssw/blob/2ba5d421e10379d81760a899532b2c991b89c82c/RecoParticleFlow/PFClusterProducer/interface/PFRecHitQTests.h#L104-L117

your comment suggests thought that some other places may have different severity thresholds applied (regular towerMaker). Is there any hit rejection done based on severity for hfreco and hbhereco rechit collections?

**Comment by abdoulline** [comment]
Slava, 

I though towerMakerPF  was used in the past (before moving to RecHits in PF), not now.
I must admit I don't exactly remember why would SeverityLevel = 11  threshold be applied anywhere.  My recollection is that there was some "softening" of this threshold in HLT configs in the past (Run1), but not in Run2...

I hope our PFlow colleagues can comment on the setting you've pointed out:   
maxSeverities = cms.vint32(11,9,9), 
And usage of the HCAL rechits (noise) flags for cleaning in general. 

For regular towerMaker the value is 9 ("regular") for all spicies of HCAL RecHits _including HF_ : 
https://cmssdt.cern.ch/lxr/source/RecoLocalCalo/CaloTowersCreator/python/calotowermaker_cfi.py#0105
And I don't see it altered anywhere in its (clones) usage, other then in aforementioned towerMakerPF:
https://cmssdt.cern.ch/lxr/search?!v=_filestring=&_string=HcalAcceptSeverityLevel&_casesensitive=1

**Comment by hatakeyamak** [comment]
There was a related discussion in recent email thread and somewhat related discussion in today's PF meeting.
https://indico.cern.ch/event/944298/

Regarding the data-specific modifications for HCAL noise flagging, most of them are now put back to the default value for eras after the phase 1 upgrade, as far as I can tell (and also as Salavat indicated above).

Only remaining data-specific modification should be:
> AddFlag(hcalRecAlgos,"HBHENegativeNoise",12) compared to MC settings of 8
> PF HBHE hits are rejected with severity > 11
I understand that this flag won't be created after HB moves to phase1 electronics in Run 3, so it won't do anything [*], but I think HCAL DPG plans to put them back to the default value explicitly starting from Run 3 (@abdoulline). Then, we no longer have data-specific modifications in Run 3 and onward.
[*]
https://cmssdt.cern.ch/lxr/source/RecoLocalCalo/HcalRecProducers/src/HBHEPhase1Reconstructor.cc#0614
https://cmssdt.cern.ch/lxr/source/RecoLocalCalo/HcalRecAlgos/interface/HBHENegativeFlag.h

For HF, we discussed a couple of optimization options for PF noise flagging for HF (these are not about data-specific customization, but simply try to see if additional noise flags will improve physics performance), and will try to follow up.

**Comment by hatakeyamak** [comment]
https://github.com/hatakeyamak/cmssw/pull/new/HFSignalAsymmetryInPF
is the branch for including HFSignalAsymmetry In PF.
@abdoulline Does this look right? I wonder if HCAL can help testing this in some collision data from Run 2. I will also ping JME.

**Comment by abdoulline** [comment]
@hatakeyamak  
I take the liberty to include Alexander @toropin as HCAL expert in noise flagging domain.

**Comment by hatakeyamak** [comment]
Thank you. @abdoulline by the way my note says:
```
Only change we want to consider is:
AddFlag(hcalRecAlgos,"HBHENegativeNoise",12) compared to MC settings of 8
PF HBHE hits are rejected with severity > 11
    → Stop doing this for Run 3, given that the HBHE phase1 electronics in Run 3. The phase1 reconstructor won’t create this flag, but still put it explicitly out of game. To be done by HCAL DPG.
```
https://docs.google.com/document/d/1tBa5q399MkSnY99cL3vILkQLj4d1h4wAgjyzfP-JSR0/edit?usp=sharing
Is this still a plan?

**Comment by abdoulline** [comment]
@hatakeyamak  Thank you for the reminder, Kenichi!

Yes, we'd need to add two lines 
     import RecoLocalCalo.HcalRecAlgos.RemoveAddSevLevel as HcalRemoveAddSevLevel
     HcalRemoveAddSevLevel.RemoveFlag(process.hcalRecAlgos,"HBHENegativeNoise")
to Run3 RECO customization right after this one:  
https://cmssdt.cern.ch/lxr/source/Configuration/DataProcessing/python/RecoTLR.py#0094
Any of us can do it at our earliest convenence,  may be you can "append" it to your submission of HFSignalAsymmetryInPF ? 

BTW, in the latter, instead of using bit 5 explictly, you may want  to 
#include "DataFormats/METReco/interface/HcalPhase1FlagLabels.h" 
and use HcalPhase1FlagLabels::HFSignalAsymmetry ?

**Comment by hatakeyamak** [comment]
> @hatakeyamak Thank you for the reminder, Kenichi!
> 
> Yes, we'd need to add two lines
> import RecoLocalCalo.HcalRecAlgos.RemoveAddSevLevel as HcalRemoveAddSevLevel
> HcalRemoveAddSevLevel.RemoveFlag(process.hcalRecAlgos,"HBHENegativeNoise")
> to Run3 RECO customization right after this one:
> https://cmssdt.cern.ch/lxr/source/Configuration/DataProcessing/python/RecoTLR.py#0094
> Any of us can do it at our earliest convenence, may be you can "append" it to your submission of HFSignalAsymmetryInPF ?

I can try. Do we need it here too:
https://cmssdt.cern.ch/lxr/source/Configuration/DataProcessing/python/RecoTLR.py#0099
I wonder if there is a more efficient way to do it for both at once.

> BTW, in the latter, instead of using bit 5 explictly, you may want to
> #include "DataFormats/METReco/interface/HcalPhase1FlagLabels.h"
> and use HcalPhase1FlagLabels::HFSignalAsymmetry ?

I see. I will think about it.

**Comment by abdoulline** [comment]
I don't think HCAL customization (formal, as no real effect is brought 
in) is explicitly needed for For Run3 "trackingOnly"... 
Slava and Joosep will correct us eventually, if needed.

Salavat

**Comment by hatakeyamak** [comment]
That makes sense. Will try to work on this this weekend.

**Comment by toropin** [comment]
I checked modifications in code to prepare PFlow objects from HF RecHits, namely,
~RecoParticleFlow/PFClusterProducer/interface/PFRecHitQTests.h and
~RecoParticleFlow/PFClusterProducer/python/particleFlowRecHitHF_cfi.py and
these modifications look fine to me.

Alexander

**Comment by hatakeyamak** [comment]
I am sorry that I couldn't respond sooner, but thank you @toropin  for checking.
I also made suggested changes by @abdoulline . I am running some local test for sanity check.
Then the next step will be to ask POGs to check this change.

**Comment by hatakeyamak** [comment]
@slava77 with this above PR, which also addressed some HCAL flag setting, can we consider this issue can be closed?

**Comment by hatakeyamak** [comment]
Pinging..

**Comment by slava77** [comment]
> @slava77 with this above PR, which also addressed some HCAL flag setting, can we consider this issue can be closed?

sorry, I wasn't able to put enough attention to this.
It's been two years since the issue was open and the full set of flags in all different options doesn't easily fit in my circular buffer memory (in part due to apparent multi-layer negation logic).


To quote from the issue description:

> Some examples of flagging of noise-like events in signal would be very useful to see if the current asymmetric approach (MC passes the hits, data rejects them) is more practical compared to a symmetric approach where both data and MC will behave the same.
> 
> The goal of this issue is for each case to see if the flag settings can be moved to the common area and apply for both data and MC.
> Additionally, some clarification is needed if relevant data about cleaned hits is saved in miniAOD (or is it available in AOD).

There are two PRs mentioned in this issue
- #29470 : seems unrelated to the main points 
- #33520 apparently added one more difference between data and MC (or did I misunderstand the change and it's actually a double-negation of something specific added for data in general in one place and then reset here for run3?)

So, it looks like we moved in the opposite direction, at least partially, away from closing this issue.

**Comment by hatakeyamak** [comment]
I believe, in #33520, `HBHENegativeNoise` was put back to the default value so that it should be the same for both data and MC in Run 3, at least eliminated one asymmetry.

Most of, if not all, data-MC asymmetry seems to be for before the hcal phase1 upgrade, and i am not sure if we can really afford studying data-MC asymmetry in detail and revisit the setting.

If the suggestion is to put various modifiers in
https://github.com/cms-sw/cmssw/blob/CMSSW_11_0_0_pre5/Configuration/StandardSequences/python/Reconstruction_Data_cff.py
https://github.com/cms-sw/cmssw/blob/CMSSW_11_0_0_pre5/Configuration/DataProcessing/python/RecoTLR.py#L47-L51
in a common place, where this "common place" should be? If that's clear, we can probably address it with guidance from @abdoulline et al.

**Comment by slava77** [comment]
Thanks for clarifying about `HBHENegativeNoise`.
what about `HFDigiTime` and `HBHEFlatNoise`?

I think that editing Reconstruction_Data_cff may be more clear (assuming the overrides in RecoTLR are removed).
It seems to me like we should be able to have the implementation more compact and clear this way.

What is the number of the removed rechits? If they are not stored already in the reduced rechit collection(s), perhaps if the size is small, some information about dropped data can be added to AOD or miniAOD.

**Comment by abdoulline** [comment]
Gentlemen, 
I don't think we're going to go back to previous ("legacy") eras to unify what was initially designed/thought as splitted (MC vs data noise filters)... 
For Run3 we have now "unified" MC & data HCAL noise flags settings...

**Comment by abdoulline** [comment]
I've checked Noise flags (SeverityLevel) settings for step3 (edmConfigDump) of wfs: 
(1) 2018 data 136.874 
(2) 2021 MC   11634.0 
and the only difference is for HBHENegativeNoise: 8(12) for MC(data).
Now, after https://github.com/cms-sw/cmssw/pull/33520  we have SeverityLevel=8 for both MC and data...

**Comment by abdoulline** [comment]
"Reduced" rechits collections in AOD contain _all_ the hits with non-0 noise flags.
And as "slimmed" collections they are available in miniAOD  https://github.com/cms-sw/cmssw/pull/31375

**Comment by hatakeyamak** [comment]
So, can we consider this is ok?

**Comment by slava77** [comment]
> So, can we consider this is ok?

I'm not sure what "this" is in this context.
I think that to close this feature more clearly it would be better to have the edits to the different flags in just one place https://github.com/cms-sw/cmssw/issues/27767#issuecomment-864089536

**Comment by hatakeyamak** [comment]
Okay. I think @abdoulline or somebody from hcal dpg will be a natural person to take care of this part, but I can assist if extra hands are helpful, as I somehow got involved..

**Comment by abdoulline** [comment]
@hatakeyamak  @slava77 
I have no reasons to object to Slava's suggestion, even though to me it seems to be more about aestetical/conveniece-wise regrouping of "scattered" re-definitions of the flags severity, than a real practical need. Well, "code reformatting" in some sense... 
But right now I have the impression (please, correct me if I'm wrong):  
(1) it's not about moving customization snippets from  RecoTLR.py to Reconstruction_Data_cff.py (historically used from the former), 
but rather 
(2) directly putting _era-specific_ HCAL data flags customization (after exising initial lines) into Reconstruction_Data_cff.py (and obviously removing those from RecoTLR_cff.py) in way similar to what's done for >=2017 both for MC and data "at source"  [*], 
just moving particular flag(s) from one group (PSet)  to the other one ?

-----------------------------------------

[*] 
https://cmssdt.cern.ch/lxr/source/RecoLocalCalo/HcalRecAlgos/python/hcalRecAlgoESProd_cfi.py#0031

**Comment by slava77** [comment]
@abdoulline 
in https://github.com/cms-sw/cmssw/issues/27767#issuecomment-864089536 I did say and mean moving the modifications from RecoTLR.py to Reconstruction_Data_cff.py
(with eras/modifiers)

**Comment by abdoulline** [comment]
@slava77 OK, thank you for confirmation about option (2). I'll try to look early next week how it could be arranged (discussing it with Kenichi @hatakeyamak, as he kindly suggested).

**Comment by abdoulline** [comment]
@hatakeyamak 
I've put together a branch with cleaned up  RecoTLR.py and extended/updated Reconstruction_Data_cff.py
https://github.com/cms-sw/cmssw/compare/master...abdoulline:Data_noise_flags_customization_regrouping

Then compared   hcalRecAlgos  content w/wo this branch in CMSSW_12_0_X_2021-06-24-1100 for 
 (1) 136.731_RunSinglePh2016B
 (2) 136.793_RunDoubleEG2017C
 (3) 136.874_RunEGamma2018C
 
(1)+(2) yiled the same content (SeverityLevel grouping) w/wo the branch,  so both look OK, 
 but (3) with the branch "loses" HFSignalAsymmetry flag from group 4: (Level=11) wrt inital setting in 
https://cmssdt.cern.ch/lxr/source/RecoLocalCalo/HcalRecAlgos/python/hcalRecAlgoESProd_cfi.py#0047
which is preserved by default for (3) without branch...
I'll try to see how it (HFSignalAsymmetry) got "lost", while 2018 era should inherit from 2017...

**Comment by abdoulline** [comment]
Follow-up: both (2) and (3) _recover_  HFSignalAsymmetry flag for Level=11 (group 4),  if I explcitly add the following fragment: 

        4 : dict( RecHitFlags = [
                'HFLongShort',
                'HFS8S1Ratio',
                'HFPET',
                'HFSignalAsymmetry']
      )

to 2017 era customization
https://github.com/cms-sw/cmssw/compare/master...abdoulline:Data_noise_flags_customization_regrouping#diff-e33077cea0c56c4aab27f549f52d82bce9ba084cc477a99800928506e65eb484R67

**Comment by hatakeyamak** [comment]
Hi @abdoulline 
Just for me to understand, why do you use:
https://github.com/cms-sw/cmssw/compare/master...abdoulline:Data_noise_flags_customization_regrouping#diff-e33077cea0c56c4aab27f549f52d82bce9ba084cc477a99800928506e65eb484R74-R86
rather than just having one line:
```
    HcalRemoveAddSevLevel.AddFlag(process.hcalRecAlgos,"HBHENegativeNoise",8)
```
?

**Comment by abdoulline** [comment]
@hatakeyamak 
if you mean two options 
https://github.com/cms-sw/cmssw/issues/27767#issuecomment-867944154
 - then Slava has confirmed (in repying to it) that option (2) with era/modifiers is the preferred one.

And once era modifiers are involved (in Reconstruction_Data_cff.py) ,  I don't see the way how to use HcalRemoveAddSevLevel.AddFlag in them ("toModify"),  but just explicitly modify parameters (PSets) of  hcalRecAlgos (ESProducer).
Please, suggest the entire (working) snippet.

I'll take this discussion "offline" to HCAL thread (with you in Cc), hoping Alexander, Igor, Jae may see something that I miss...

**Comment by abdoulline** [comment]
@slava77   Slava, in view of Kenichi's doubts (see above https://github.com/cms-sw/cmssw/issues/27767#issuecomment-870028689  ), could you, please confirm once again that what you meant in your short answer 
https://github.com/cms-sw/cmssw/issues/27767#issuecomment-867976054
corresponds to the tentative modifications in the branch :
https://github.com/cms-sw/cmssw/compare/master...abdoulline:Data_noise_flags_customization_regrouping

**Comment by slava77** [comment]
> @slava77 Slava, in view of Kenichi's doubts (see above [#27767 (comment)](https://github.com/cms-sw/cmssw/issues/27767#issuecomment-870028689) ), could you, please confirm once again that what you meant in your laconic answer
> [#27767 (comment)](https://github.com/cms-sw/cmssw/issues/27767#issuecomment-867976054)
> corresponds to the tentative modifications in the branch :
> [master...abdoulline:Data_noise_flags_customization_regrouping](https://github.com/cms-sw/cmssw/compare/master...abdoulline:Data_noise_flags_customization_regrouping)

I can confirm my comment.
I would like to clarify that that comment implied that the implementation is clear and does not blow up to almost 100 lines.
It should be possible to call `HcalRemoveAddSevLevel.AddFlag(hcalRecAlgos` for each specific modifier with smth like
```python
def _modName(algos):
   HcalRemoveAddSevLevel.AddFlag(algos,"HBHEFlatNoise",12)
someModifier.toModify(hcalRecAlgos, _modName)
```

**Comment by abdoulline** [comment]
@slava77  thank you for the suggestion.
I've admitted earlier (and I must re-admit) I didn't know this rather simple way of era-related executon of particuar algo. 
I'll try to follow it.

**Comment by abdoulline** [comment]
@slava77   new branch:
https://github.com/cms-sw/cmssw/compare/master...abdoulline:Data_noise_flags_move
seems to work fine (SeverityLevel groups are identical w/wo branch)  for 
(1) 136.731_RunSinglePh2016B
(2) 136.793_RunDoubleEG2017C
(3) 136.874_RunEGamma2018C
I'll submit PR later today (and after passing via runTheMatrix tests)...

**Comment by slava77** [comment]
>  I'll submit PR later today (after passing via runTheMatrix tests)...

thanks.
- Please group the modifications for one era in one method
- please also add some notes clarifying (where applicable) that the applied modifications are reverting the changes back to the default used for MC

**Comment by abdoulline** [comment]
<...>Please group the modifications for one era in one method<...>
I'm afraid I don't quite understand what you mean?  - Something like this for the first run2_25ns_specific case with 2 modifications: 
https://github.com/cms-sw/cmssw/compare/master...abdoulline:Data_noise_flags_move#diff-e33077cea0c56c4aab27f549f52d82bce9ba084cc477a99800928506e65eb484R28
def _modName(algos):
   HcalRemoveAddSevLevel.AddFlag(algos,"HBHEFlatNoise",8)
   HcalRemoveAddSevLevel.AddFlag(algos,"HFDigiTime",8)

?

**Comment by slava77** [comment]
> something like this for the first run2_25ns_specific case with 2 modifications: ...

yes

**Comment by abdoulline** [comment]
OK, thanks!

**Comment by hatakeyamak** [comment]
Thank you @abdoulline

**Comment by hatakeyamak** [comment]
Thank you @abdoulline.
@slava77 perhaps now we can consider this is concluded?

**Comment by slava77** [comment]
> Thank you @abdoulline.
> @slava77 perhaps now we can consider this is concluded?

I agree, we can close this issue, it is resolved with #34282.

**Comment by slava77** [comment]
+1
